statistics/handle/statscache.go

```diff
@@ -15,15 +15,274 @@ package handle
 
 import (
 	"encoding/binary"
+	"errors"
 	"sync"
+	"time"
 
+	"github.com/pingcap/badger/cache"
 	"github.com/pingcap/tidb/statistics"
 	"github.com/pingcap/tidb/util/kvcache"
 	"github.com/pingcap/tidb/util/memory"
 )
 
-// statsCache caches table statistics.
-type statsCache struct {
+// StatsCache a interface can LookUp Update
+type StatsCache interface {
+	Lookup(id int64) (*statistics.Table, bool)
+	Update(tables []*statistics.Table, deletedIDs []int64, newVersion uint64)
+	GetVersion() uint64
+	InitStatsCache(tables map[int64]*statistics.Table, version uint64)
+	GetAll() []*statistics.Table
+
+	// interface below are used only for test
+	Clear()
+	Close()
+	GetBytesLimit() int64
+	SetBytesLimit(bytesLimit int64)
+	BytesConsumed() int64
+}
+
+const byteShards = 8
+const numShards = 1 << byteShards
+
+// statsCache caches Regions loaded from PD.
+type ristrettoStatsCache struct {
+	mu         sync.Mutex
+	cache      *cache.Cache
+	version    uint64
+	memTracker *memory.Tracker // track memory usage.
+	tablesMap
+}
+
+type tablesMap struct {
+	tablesShards [numShards]tableShard
+}
+
+// tableShard is a shard of table
+type tableShard struct {
+	data map[int64]*statistics.Table
+	sync.RWMutex
+}
+
+// Set set key with value
+func (ts *tablesMap) Set(key int64, value *statistics.Table) {
```

Comment:
Nitpick: tablesMap is part of ristrettoStatsCache, so Set/Get/Del shouldn't be exported..